stl/inc/xstring

```diff
@@ -2589,19 +2598,26 @@ public:
     }
 
 private:
-    static constexpr void _Start_element_lifetimes(_Elem* const _Ptr, const size_type _Size) {
+    static _CONSTEXPR20 pointer _Allocate_from_al(_Alty& _Al, const size_type _New_capacity) {
+        _STL_INTERNAL_CHECK(_New_capacity > _SMALL_STRING_CAPACITY);
+        const pointer _Fancy_ptr = _Al.allocate(_New_capacity + 1); // throws
         // Start element lifetimes to avoid UB. This is a more general mechanism than _String_val::_Activate_SSO_buffer,
         // but likely more impactful to throughput.
 #if _HAS_CXX20
         if (_STD is_constant_evaluated()) {
-            for (size_type _Idx = 0; _Idx < _Size; ++_Idx) {
+            _Elem* const _Ptr = _Unfancy(_Fancy_ptr);
+            for (size_type _Idx = 0; _Idx < _New_capacity + 1; ++_Idx) {
                 _STD construct_at(_Ptr + _Idx);
             }
         }
-#else // ^^^ C++20-or-later / pre-C++20 vvv
-        (void) _Ptr;
-        (void) _Size;
 #endif // _HAS_CXX20
+        return _Fancy_ptr;
+    }
+
+    static _CONSTEXPR20 void _Deallocate_from_al(
+        _Alty& _Al, const pointer _Old_ptr, const size_type _Old_capacity) noexcept {
+        _STL_INTERNAL_CHECK(_Old_capacity > _SMALL_STRING_CAPACITY);
+        _Al.deallocate(_Old_ptr, _Old_capacity + 1);

```

Comment:
I'm not too confident about this one. Is passing `pointer` by value ok here?

Reply:
> I'm not too confident about this one. Is passing `pointer` by value ok here?

IMO yes. If someone provides a `pointer` type which is not OK for passing by value, that the allocator won't meet the _Cpp17Allocator_ named requirements and thus the behavior is undefined.

Reply:
My concern is that, `pointer` looks not necessarily trivially copyable, so it's possible that this call will not be an exact replacement to direct `_Al.deallocate` call in some rare cases.